### PR TITLE
Refactored + added close method

### PR DESCRIPTION
**Pull request by lukeburns**
Added close method for https://github.com/karissa/hyperdrive-archive-swarm/issues/2

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,78 +1,96 @@
...
-        if (err) errors++
-        if (--pending) return
-        if (errors === urls.length) return cb(err)
-        cb()
-      })
+  var pending = this.urls.length
+  var errors = 0
+
+  var self = this
+  return this.urls.forEach(function (url) {
```
Comment (mafintosh) [inline index.js]: why is there a return statement here?

File: index.js [hunk 2]
```diff
@@ -1,78 +1,96 @@
...
+    var bc = broadcast(url, channel, message, function (err) {
```
Comment (mafintosh) [inline index.js]: is `bc` being used anywhere?

File: index.js [hunk 3]
```diff
@@ -1,78 +1,96 @@
...
-      url: url + '/v1/' + app + '/' + channel
-    }, function (err, res) {
-      if (err) return cb(err)
-      if (res.statusCode !== 200) return cb(new Error('Bad status: ' + res.statusCode))
-      cb()
-    })
-  }
+  this.on('close', function (cb) {
+    all.destroy()
+  })
```
Comment (mafintosh) [inline index.js]: this listener should be returned on `all.on('close')`

File: index.js [hunk 4]
```diff
@@ -1,78 +1,96 @@
...
+inherits(SignalHub, events.EventEmitter)
 
-    if (streams.length === 1) return streams[0]
+SignalHub.prototype.subscribe = function (channel) {
+  var endpoint
+  if (Array.isArray(channel)) {
+    endpoint = channel.join(',')
+  } else {
+    endpoint = channel
+  }
```
Comment (mafintosh) [inline index.js]: since we're refactoring we might as well turn this into a ternary statement

File: index.js (hunk 3 above)
Comment (mafintosh) [inline index.js]: and we probably need this.setMaxListeners(0)

File: index.js [hunk 5]
```diff
@@ -1,78 +1,96 @@
...
+      cb()
     })
-  }
+  })
+}
 
-  return that
+SignalHub.prototype.close = SignalHub.prototype.destroy = function (cb) {
+  this.emit('close')
+  if (typeof cb === 'function') cb()
```
Comment (mafintosh) [inline index.js]: technically it would be better to wait for all subscribers to close before emitting close but i'm ok with this for now.

**Comment by mafintosh** [comment]
added some comments. could you document this method in the readme? also i'm ok with just calling this .close and not supporting .destroy but thats up to you

File: index.js (hunk 1 above)
Comment (lukeburns) [inline index.js]: For no good reason

File: index.js (hunk 2 above)
Comment (lukeburns) [inline index.js]: Nope got rid of it

File: index.js (hunk 3 above)
Comment (lukeburns) [inline index.js]: I setMaxListeners(0).

File: index.js (hunk 4 above)
Comment (lukeburns) [inline index.js]: 👍

File: index.js (hunk 5 above)
Comment (lukeburns) [inline index.js]: If I close a pump stream, does it close its members?

**Comment by lukeburns** [comment]
I think I addressed all of your comments in this commit. Is this how you were imagining the close method?

**Comment by lukeburns** [comment]
What's a good way of waiting for subscribers to close before emitting a close event?

File: index.js [hunk 6]
```diff
@@ -1,78 +1,106 @@
...
+  })
+}
 
-    urls.forEach(function (url) {
-      broadcast(url, channel, message, function (err) {
-        if (err) errors++
-        if (--pending) return
-        if (errors === urls.length) return cb(err)
-        cb()
+SignalHub.prototype.close = function (cb) {
```
Comment (mafintosh) [inline index.js]: you can add `if (cb) this.once('close', cb)` and then get rid of the cb calls below

File: index.js [hunk 7]
```diff
@@ -1,78 +1,106 @@
...
-      method: 'POST',
-      json: message,
-      url: url + '/v1/' + app + '/' + channel
-    }, function (err, res) {
-      if (err) return cb(err)
-      if (res.statusCode !== 200) return cb(new Error('Bad status: ' + res.statusCode))
-      cb()
-    })
-  }
+  this.subscribers.push(subscriber)
```
Comment (mafintosh) [inline index.js]:
``` js
var self = this
subscriber.once('close', function () {
  var i = self.subscribers.indexOf(subscriber)
  if (i > -1) self.subscribers.splice(i, 1)
})
```

**Comment by mafintosh** [comment]
Made it throw if .broadcast / .subscribe was called after close. Out in 4.7.0

**Pull request merged**